PYIC-7507: Journey routing for DL auth source checks

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed
Journey routing for DL auth source checks

### Why did it change

This adds journey routing to all the instances of a DCMAW state we have, and handles their unique subtleties... sort of.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-7507](https://govukverify.atlassian.net/browse/PYIC-7507)


[PYIC-7507]: https://govukverify.atlassian.net/browse/PYIC-7507?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-7507: Journey routing for DL auth source checks

This adds journey routing to all the instances of a DCMAW state we have,
and handles their unique subtleties... sort of.

Wynndow commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml:
```diff
@@ -730,6 +801,16 @@ states:
+      next:
+        targetJourney: EVALUATE_SCORES
+        targetState: START
+
```
This is for same session enhanced verification mitigation. If the user backs out of the drivingLicence CRI  we'll get an access-denied event. At written this will fail the users journey. But we could instead make changes to the reset-session-identity lambda to clear out the DCMAW VC and offer the user F2F as another way to achieve a verification score.
We could potentially do this for any non CI attracting failure from driving licence.

Joe-Edwards-GDS replied:
Ah, so this is the specific case where the user has used a passport in the web route, but for some reason chose to use a DL for the DCMAW mitigation?

We can take it to Joe/Simon to confirm, but I think it's probably fine to let them fail at that point.

Wynndow replied:
Yep fair - anything that isn't a `next` event here will fallback to the defailt parent states, which are all fails.

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/f2f-hand-off.yaml:
```diff
@@ -45,6 +45,9 @@ states:
         targetState: FAILED
+      dl-auth-source-check:
+        targetJourney: FAILED
+        targetState: FAILED
```
It's a bit annoying that we now have to add this as a state for _all_ CRIs - even ones that will never emit it :/

I don't think it's _that_ bad - and does ensure that we never get an unhandled state :shrug:. Otherwise we could finesse the journey map test so that it's aware of certain CRI-specific cases.

Wynndow replied:
Yeah it is a little annoying, but it does give us some protection for future changes. I wonder if we could define parent states in one place, and then have other journey maps reference them...

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p1-identity.yaml:
```diff
@@ -79,6 +79,9 @@ states:
         targetState: FAILED
+      dl-auth-source-check:
+        targetJourney: FAILED
+        targetState: FAILED
```
Do you think `FAILED` or `TECHNICAL_ERROR` is a better default? (Since this event is still a 'success')

Of course it's a largely academic question!

Wynndow replied:
I suppose technical error may be better, as the user will have gotten themselves into an unexpected situation. I think I was following the pattern for the events above which all come into the same category. I'll this one to technical error, but might be worth applying that logic across the board.

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p1-identity.yaml:
```diff
@@ -198,6 +203,41 @@ states:
+  CRI_DRIVING_LICENCE_AUTH_SOURCE_CHECK:
+    response:
+      type: cri
+      criId: drivingLicence
```
As per https://gds.slack.com/archives/C068GQY76EL/p1727363487409729 I think we'll want to add `context: check_details` here as well

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml:
```diff
@@ -634,6 +676,33 @@ states:
             targetJourney: INELIGIBLE
             targetState: INELIGIBLE
 
+  MITIGATION_01_CRI_DRIVING_LICENCE_AUTH_SOURCE_CHECK:
```
All these wordy states are a bit awkward in the journey map but I don't have a better suggestion 🤷

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/new-p2-identity.yaml:
```diff
@@ -634,6 +676,33 @@ states:
+      access-denied:
+        targetState: MITIGATION_01_RESET_SESSION_BEFORE_PYI_POST_OFFICE
+      temporarily-unavailable:
+        targetState: MITIGATION_01_RESET_SESSION_BEFORE_PYI_POST_OFFICE
```
Why do we map `temporarily-unavailable` here, but not for any of the others?

Wynndow replied:
Do you mean the other events that are defined on `MITIGATION_01_CRI_DCMAW`? If so the other events are `not-found`, `fail-with-no-ci` and `enhanced-verification` - none of which should be possible to get from the DL CRI I think?

Joe-Edwards-GDS replied:
I mean we don't handle `temporarily-unavailable` on any other instances of the driving licence CRI

(afaict we only handle it from DCMAW - so I assume other CRIs never actually emit it, but happy to be proved wrong!)

Wynndow replied:
I figured we're handling it on `MITIGATION_01_CRI_DCMAW`, so by extension to preserve the functionality of the system (offer them a f2f mitigation route if they can't do it with the app), we'd need to handle it here.
Thinking about it though, this is part of the wider conversation about mitigations. I'll route this to a failure state.

Joe-Edwards-GDS reviewed (commented):
A couple of thoughts:
- We haven't properly defined the escape routes yet - I think you've defined some sensible options, but they may change.
- Is there any value in extracting DCMAW into a nested journey, now that it's potentially more than just a single state?
    - Could also be a separate task - it needs to be a two-step change to avoid breaking existing states

Wynndow committed:
PYIC-7507: Add check_details context to auth source check states

The DL CRI need to know if the user is coming to them for a journey that
requires them to check a document, or if it's a journey that's to just
check already gathered document details. This will be established
through the presence of a check_details context claim in the top level
of the jar.

Wynndow committed:
PYIC-7507: Set default journey to technical error

This updates the default journey for a dl-auth-source-check event on the
CRI parent states to a technical error. If a user ever receives this
event in a place we don't explicitly handle it, this would be a
technical error.

Wynndow committed:
PYIC-7507: Route to technical error on temp unavailable

If the DL CRI is unavailable, we should probably just route to a
technical failure for now. There is a wider concern around mitigations,
but that will be picked up in the future.

Wynndow committed:
PYIC-7505: API tests for DL auth source check journeys

Adds a feature file to test our journeys with the auth source check flag
enabled. Once the feature is releases the tests can be subsumed by the
other feature files.

Wynndow commented on api-tests/features/dl-auth-source-checks.feature:
```diff
@@ -0,0 +1,183 @@
+    When I submit 'kenneth-driving-permit-needs-alternate-doc' details with attributes to the CRI stub
+      | Attribute | Values          |
+      | context   | "check_details" |
+    Then I get a 'pyi-driving-licence-no-match-another-way' page response
```
It would be nice if we could test that the reset-session lambda is run at this point, but we don't have a mechanism, and the lambda doesn't emit an audit event...

Joe-Edwards-GDS replied:
I guess the 'test' would need to complete the mitigation journey and either:
- check that the identity bundle at the end doesn't contain a DCMAW VC
- complete the mitigation route with a different name (on the basis that it would fail correlation if the DCMAW VC wasn't wiped) - but this is a pretty indirect way of doing it!

Wynndow replied:
Ah yeah, option 1 is the simplest. I'll add it in.

Joe-Edwards-GDS commented on api-tests/features/dl-auth-source-checks.feature:
```diff
@@ -0,0 +1,183 @@
+      | ukPassport | kenneth-passport-valid              |
+      | address    | kenneth-current                     |
+      | fraud      | kenneth-score-2                     |
+      | kbv        | kenneth-needs-enhanced-verification |
```
One observation about this - with the session store, you actually won't have any of these credentials - you'll just have the CI.

It's not really a problem (and historic users might well have the credentials, at least until we perform the EVCS migration).

Wynndow replied:
Yeah, that's a good point - sort of means the changes I recently made to post to cimit in this step are a bit off. I'll look at fixing it up and adding something to just send CIs to cimit, but probably not in this PR

Joe-Edwards-GDS reviewed (dismissed):
A couple of notes about the tests but otherwise looks good!

We can think about extracting a nested journey for DCMAW after this, but could be easier once we have the full scope.

One last thought - do you think there's any value in putting the journey routes behind a feature flag? I can't really see one, since generation of the event is feature-flagged instead...

Wynndow committed:
PYIC-7507: Extend test to ensure no DL VC

This makes sure that the reset lambda is run in the background.

Also extends the existing test to be more flexible, allowing us to check
for the presence of absence of any VC in the identity.

Joe-Edwards-GDS commented on api-tests/features/dl-auth-source-checks.feature:
```diff
@@ -31,6 +31,10 @@ Feature: Authoritative source checks with driving licence CRI
+    When I submit an 'end' event
+    Then I get an OAuth response
+    When I use the OAuth response to get my identity
+    Then I get a 'P0' identity without a drivingLicence VC
```
Do you mean `drivingLicence` or `dcmaw` here? I think dcmaw is the problematic one, since it'll be valid!

Wynndow replied:
oh flip, good spot!

Joe-Edwards-GDS reviewed (approved):
Thanks!
